src/Security/BasicAuthMiddleware.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+namespace SilverStripe\Security;
+
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Control\HTTPResponse_Exception;
+use SilverStripe\Control\Middleware\HTTPMiddleware;
+use SilverStripe\Control\Middleware\CanonicalURLMiddleware;
+
+class BasicAuthMiddleware implements HTTPMiddleware
+{
+    /**
+     * URL patterns that the global protection should not be applied to.
+     * Does nothing if SilverStripe\Security\BasicAuth.entire_site_protected is false.
+     *
+     * Not affected by IncludeURLPatterns
+     *
+     * @see CanonicalURLMiddleware
+     * @var array
+     */
+    protected $excludeURLPatterns = [];
+
+    /**
+     * URL to include, even if entire_site_protected is false.
+     * The list can either be an array of patterns to check, or an associative array
+     * of patterns mapped to either a permission code, or array of codes to check.
+     *
```

Comment:
This is a bit of a complex spec to describe with written docs only. As a dev using this for the first time, I'd much appreciate an example here, e.g.

```
[
  'my-secret-page' => 'ADMIN',
  '/^members-only\/*$',
]
```

Reply:
I've simplified this to a single array mapping patterns to rules.